### PR TITLE
feat: tweak launcher to launch lopsided jobs

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change tweaks the distributed training launchers to accommodate jobs with different sized agents.

## Test Plan
- [x] spin up a cluster with 1 1xGPU node and 1 2xGPU node
- [x] modify pytorch distributed, horovod and deepspeed experiments to run
- [x] ensure they launch ok
- [x] for release party, just ping brad and borrow his cluster, since this is such a pain to test.

## Commentary (optional)
There were a few subtle bugs in the ipc code this revealed (things that weren't technically correct, but always worked if nodes were homogeneous).

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/launch/deepspeed.py [hunk 1]
```diff
@@ -73,7 +75,11 @@ def create_hostlist_file(
     if hostfile_path is not None and not os.path.exists(hostfile_path):
         os.makedirs(os.path.dirname(hostfile_path), exist_ok=True)
         with open(hostfile_path, "w") as hostfile:
-            lines = [f"{host} slots={num_proc_per_machine}\n" for host in trial_runner_hosts]
+
```
Comment (rb-determined-ai) [inline harness/determined/launch/deepspeed.py:78]: nit: this is a weird place to put an empty line

File: harness/determined/exec/prep_container.py [hunk 2]
```diff
@@ -106,11 +114,17 @@ def do_rendezvous_slurm(
             data={
                 "rank": rank,
                 "rendezvous_ip": rendezvous_ip,
+                "slots": num_slots,
             },
         ),
     )
     addrs = [d["rendezvous_ip"] for d in sorted(resp.data, key=lambda d: int(d["rank"]))]
-    return det.RendezvousInfo(container_addrs=addrs, container_rank=rank)
+    slots = [d["slots"] for d in sorted(resp.data, key=lambda d: int(d["rank"]))]
```
Comment (rb-determined-ai) [inline harness/determined/exec/prep_container.py:122]: nit: sort once and reuse the result?

File: harness/determined/_info.py [hunk 3]
```diff
@@ -353,6 +359,16 @@ def container_addrs(self) -> List[str]:
         assert self._rendezvous_info is not None
         return self._rendezvous_info.container_addrs
 
+    @property
+    def container_slots(self) -> List[int]:
```
Comment (rb-determined-ai) [inline harness/determined/_info.py:363]: Can we rename this?  I don't think the name clearly reflects the meaning.

If I'm on an 4-node, 2-gpu-per-node job, I would have imagined `container_slots` to be `2`, not `[2,2,2,2]`.

Maybe `container_slot_counts`?  My thinking is that since "slots" is naturally something that is already plural on every node, but the "slot_count" can only be one per node, so the plural "slot_counts" naturally sounds more like it should be a list of slot counts per container.

File: harness/determined/horovod.py [hunk 4]
```diff
@@ -85,11 +85,15 @@ def cross_size(self) -> Any:
 hvd = _PolyHorovod()
 
 
-def create_hostlist_arg(num_proc_per_machine: int, ip_addresses: List[str]) -> str:
+def create_hostlist_arg(num_proc_per_machine: List[int], ip_addresses: List[str]) -> str:
```
Comment (rb-determined-ai) [inline harness/determined/horovod.py:88]: can you rename `num_proc_per_machine` to be `slot_counts` or `container_slot_counts` or `per_node_slot_counts` or something?  I dislike that it used to mean the same thing as torch.distributed.launch's `--num_procs_per_node` but now means something entirely different.

I also think that the current name reflects the old meaning well, but is a bit confusing wit the new meaning.

File: harness/tests/test_horovod.py [hunk 5]
```diff
@@ -70,7 +70,7 @@ def test_create_run_command(
 
 def test_create_hostlist_arg() -> None:
     ip_addresses = ["localhost", "128.140.2.4"]
-    num_proc_per_machine = 8
+    num_proc_per_machine = [8, 8]
     expected_horovod_hostlist_arg = (
         f"{ip_addresses[0]}:{num_proc_per_machine},{ip_addresses[1]}:{num_proc_per_machine}"
     )
```
Comment (rb-determined-ai) [inline harness/tests/test_horovod.py:76]: this looks like it couldn't possibly pass.

Also I'd recommend just hardcoding the expect string as "localhost:8,128.140.2.4:8" which would be easier to read anyway.

**Comment by rb-determined-ai** [comment]
So according to @garrett361, pytorch totally does [actually support lopsided jobs](https://discuss.pytorch.org/t/can-i-have-different-number-of-gpus-on-different-nodes-while-running-distributed-tasks/156709/6)

**Review by rb-determined-ai — APPROVED** [review]
Python side looks good, really nice work.

File: harness/determined/launch/torch_distributed.py [hunk 6]
```diff
@@ -75,6 +75,8 @@ def main(override_args: List[str], script: List[str]) -> int:
     info = det.get_cluster_info()
     assert info is not None, "must be run on-cluster"
 
+    assert len(set(info.container_slots)) == 1, "pytorch dtrain doesn't support lopsided jobs"
```
Comment (stoksc) [inline harness/determined/launch/torch_distributed.py:78]:
```suggestion
    assert len(set(info.container_slots)) == 1, "Determined doesn't support lopsided jobs in pytorch native dtrain"
```

File: harness/tests/test_horovod.py [hunk 7]
```diff
@@ -69,12 +69,6 @@ def test_create_run_command(
...
 def test_create_hostlist_arg() -> None:
-    ip_addresses = ["localhost", "128.140.2.4"]
-    num_proc_per_machine = 8
-    expected_horovod_hostlist_arg = (
-        f"{ip_addresses[0]}:{num_proc_per_machine},{ip_addresses[1]}:{num_proc_per_machine}"
-    )
-    created_horovod_hostlist_arg = horovod.create_hostlist_arg(
-        num_proc_per_machine=num_proc_per_machine, ip_addresses=ip_addresses
+    assert "localhost:8,128.140.2.4" == horovod.create_hostlist_arg(
+        host_slot_counts=["localhost", "128.140.2.4"], ip_addresses=[8, 8]
```
Comment (stoksc) [inline harness/tests/test_horovod.py:73]:
```suggestion
        ip_addresses=["localhost", "128.140.2.4"], host_slot_counts=[8, 8]
```

File: harness/determined/core/_distributed.py [hunk 8]
```diff
@@ -68,7 +68,7 @@ def __init__(
         self._is_chief = self.rank == 0
         self._is_local_chief = self.local_rank == 0
 
-        if self.cross_size > 1:
+        if self.local_size != self.size:
```
Comment (mpkouznetsov) [inline harness/determined/core/_distributed.py:71]: The AssertionError should maybe say "This a distributed job (`local_size != size`) but chief_ip was not provided ..."

**Review by mpkouznetsov — APPROVED** [review]
One suggestion for an (optional) error message tweak, otherwise LGTM.

File: harness/determined/core/_distributed.py [hunk 9]
```diff
@@ -68,7 +68,7 @@ def __init__(
...
             if chief_ip is None:
                 raise AssertionError(
                     f"rank_info has cross_size ({self.cross_size}) but chief_ip was not "
```
Comment (stoksc) [inline harness/determined/core/_distributed.py:74]:
```suggestion
                    f"rank_info has self.local_size != self.size ({self.local_size} != {self.size}) but chief_ip was none"
```

File: harness/determined/core/_distributed.py [hunk 10]
```diff
@@ -116,7 +116,11 @@ def _init_ipc(self, force_tcp: bool) -> None:
 
         # Local broadcast server.
         self.tempdir = None
-        if self._is_local_chief:
+        if self.local_size < 2:
+            # If local size is less than 2, we don't need a local chief but still need to parcipate
```
Comment (stoksc) [inline harness/determined/core/_distributed.py:120]:
```suggestion
            # If local size is less than 2, we don't need a local chief but still need to participate
```

File: harness/determined/core/_distributed.py [hunk 11]
```diff
@@ -68,10 +68,10 @@ def __init__(
         self._is_chief = self.rank == 0
         self._is_local_chief = self.local_rank == 0
 
-        if self.cross_size > 1:
+        if self.local_size != self.size:
             if chief_ip is None:
                 raise AssertionError(
-                    f"rank_info has cross_size ({self.cross_size}) but chief_ip was not "
+                    f"rank_info has self.local_size != self.size ({self.local_size} != {self.size}) but chief_ip was none"
```
Comment (stoksc) [inline harness/determined/core/_distributed.py:74]:
```suggestion
                    f"This is a distributed job (local_size {self.local_size} != size {self.size}) but chief_ip was none"
```
